### ISSUE TITLE
False Negative: Precedence-sensitive expressions stop triggering once they become slightly more complex.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5641]
# False Negative: Precedence-sensitive expressions stop triggering once they become slightly more complex.

Version
2.48.0

## Checker
- Checker id: `OperatorPrecedence`
- Checker description: This checker detects binary expressions or conditional expressions where operator precedence is potentially confusing and suggests adding parentheses for clarity.

## Description of the false negative
`diff/output` marks 3 false negatives for `OperatorPrecedence`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are string reference equality.

### `PosCase2.java`
Source: `output/sonarqube/OperatorPrecedenceCheck/src/main/java/scensct/core/pos/PosCase2.java`

The core operation is a reference-equality check involving strings. The misuse is still a plain reference comparison on string values, which should not depend on whether one operand is a literal or a variable. That is also how the benchmark classifies the sample: If statement condition is assignment where RHS is binary relational expression should be flagged as ambiguous.

```java
// If statement condition is assignment where RHS is binary relational expression should be flagged as ambiguous.
package scensct.core.pos;

public class PosCase2 {
    public static void main(String[] args) {
        boolean flag;
        int x = 5, y = 10;
        // Scenario 2: Assignment = inside if with relational == on RHS causes precedence ambiguity.
        if (flag = x == y) { // Ambiguous: == has higher precedence than =, parentheses suggested. // [REPORTED LINE]
            System.out.println("Assigned and compared");
        }
    }
}
```

### `PosCase3.java`
Source: `output/sonarqube/OperatorPrecedenceCheck/src/main/java/scensct/core/pos/PosCase3.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Conditional expression where trueExpression is complex (binary expression) should be flagged as needing parentheses.

```java
// Conditional expression where trueExpression is complex (binary expression) should be flagged as needing parentheses.
package scensct.core.pos;

public class PosCase3 {
    public static void main(String[] args) {
        int a = 5, b = 10, c = 15;
        // Scenario 3: trueExpression is a + b * c (binary expression), which is complex.
        int result = (a > 0) ? a + b * c : 1; // trueExpression is complex, should be parenthesized. // [REPORTED LINE]
        System.out.println(result);
    }
}
```

### `PosCase4.java`
Source: `output/sonarqube/OperatorPrecedenceCheck/src/main/java/scensct/core/pos/PosCase4.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Conditional expression where falseExpression is complex (binary expression) should be flagged as needing parentheses.

```java
// Conditional expression where falseExpression is complex (binary expression) should be flagged as needing parentheses.
package scensct.core.pos;

public class PosCase4 {
    public static void main(String[] args) {
        int a = 5, b = 10, c = 15;
        // Scenario 4: falseExpression is a - b / c (binary expression), which is complex.
        int result = (a > 0) ? 1 : a - b / c; // falseExpression is complex, should be parenthesized. // [REPORTED LINE]
        System.out.println(result);
    }
}
```

## Cause analysis
The rule is documented as follows: This checker detects binary expressions or conditional expressions where operator precedence is potentially confusing and suggests adding parentheses for clarity.

In this set the problem shows up around string reference equality. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `OperatorPrecedence`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.
